newrelic/hooks/database_psycopg.py

```diff
@@ -0,0 +1,500 @@
+                cursor = await self.__wrapped__.execute(sql, **kwargs)
+
+        return self.__cursor_wrapper__(cursor, self._nr_dbapi2_module, self._nr_connect_params, (args, kwargs))
+
+    async def __aenter__(self):
+        name = callable_name(self.__wrapped__.__aenter__)
+        with FunctionTrace(name, source=self.__wrapped__.__aenter__):
+            await self.__wrapped__.__aenter__()
+
+        # Must return a reference to self as otherwise will be
+        # returning the inner connection object. If 'as' is used
+        # with the 'with' statement this will mean no longer
+        # using the wrapped connection object and nothing will be
+        # tracked.
+
+        return self
+
+    async def __aexit__(self, exc, value, tb):
+        name = callable_name(self.__wrapped__.__aexit__)
+        with FunctionTrace(name, source=self.__wrapped__.__aexit__):
+            if exc is None:
+                with DatabaseTrace(
+                    sql="COMMIT",
+                    dbapi2_module=self._nr_dbapi2_module,
+                    connect_params=self._nr_connect_params,
+                    source=self.__wrapped__.__aexit__,
+                ):
+                    return await self.__wrapped__.__aexit__(exc, value, tb)
+            else:
+                with DatabaseTrace(
+                    sql="ROLLBACK",
+                    dbapi2_module=self._nr_dbapi2_module,
+                    connect_params=self._nr_connect_params,
+                    source=self.__wrapped__.__aexit__,
+                ):
+                    return await self.__wrapped__.__aexit__(exc, value, tb)
+
+
+# This connection wrapper does not save cursor parameters for explain plans. It
+# is only used for the default connection class.
+class AsyncConnectionWrapper(AsyncConnectionSaveParamsWrapper):
+    def cursor(self, *args, **kwargs):
+        # If any unknown cursor params are detected or a cursor factory is
+        # used, store params for explain plans later.
+        if should_preserve_cursor_args(*args, **kwargs):
+            cursor_params = (args, kwargs)
+        else:
+            cursor_params = None
+
+        return self.__cursor_wrapper__(
+            self.__wrapped__.cursor(*args, **kwargs), self._nr_dbapi2_module, self._nr_connect_params, cursor_params
+        )
+
+
+class AsyncConnectionFactory(DBAPI2AsyncConnectionFactory):
+
+    __connection_wrapper__ = AsyncConnectionWrapper
+
+    async def __call__(self, *args, **kwargs):
+        if should_preserve_connection_args(self, *args, **kwargs):
+            self.__connection_wrapper__ = AsyncConnectionSaveParamsWrapper
+
+        return await super(AsyncConnectionFactory, self).__call__(*args, **kwargs)
+
+
+def instance_info(args, kwargs):
+
+    p_host, p_hostaddr, p_port, p_dbname = _parse_connect_params(args, kwargs)
+    host, port, db_name = _add_defaults(p_host, p_hostaddr, p_port, p_dbname)
+
+    return (host, port, db_name)
+
+
+def _parse_connect_params(args, kwargs):
```

Comment:
Part of me wants to move these helper methods out of here and put them in a shared location. I realize that one is ever so slightly different. I wonder if we could just abstract that difference to an additional input argument or something similar. It would be easier for comparison purposes between the two versions and less code duplication.

Reply:
One of the problems is importing other hooks files will report that instrumentation package as used by the environment reporting feature, which just checks sys.modules for things. If we don't put it in another hooks file we'll be putting code specific to a few instrumentation packages in common places, which messes with project structure. 

I personally feel inclined to just let duplication happen, so that if we change something in this code for psycopg3 it doesn't possibly break psycopg2 in the future. We've gotten bit pretty hard in the past with trying too hard to reuse code an avoid duplication. (See everything async redis related.)